tests/thirdparty/test_tensorflow.py

```diff
@@ -0,0 +1,39 @@
+# Copyright 2020 The Trieste Contributors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import pytest
+import tensorflow as tf
+
+
+def test_raise_is_incompatible_with_tf_function() -> None:
+    @tf.function
+    def f(a: tf.Tensor) -> tf.Tensor:
+        if a <= tf.constant(0):
+            raise ValueError
+
+        return a
+
+    with pytest.raises(ValueError):
+        f(tf.constant(1))  # note that 1 should *not* trigger the error branch, but does
```

Comment:
i'm surprised by this, so it would be good if someone checked i've got this right

Reply:
this seems more line the problem is in the if-else statement rather than raising exceptions?

Reply:
why "thirdparty"?

Reply:
> this seems more line the problem is in the if-else statement rather than raising exceptions?
the if-else is fine. `tf.function` executes every branch of a function when tracing, but it apparently doesn't compile exceptions into the graph, it just raises them.

> why "thirdparty"?
"thirdparty" cos we're testing tensorflow rather than trieste